DB/src/main/java/io/deephaven/db/util/PythonScopeJpyImpl.java

```diff
@@ -67,6 +67,72 @@ public CallableWrapper(PyObject pyObject) {
         public Object call(Object... args) {
             return convert(pyObject.callMethod("__call__", args));
         }
+
+        public PyObject getPyObject() {
+            return pyObject;
+        }
+    }
+
+    public static final class NumbaCallableWrapper extends CallableWrapper {
+        private List<Class> paramTypes;
+        private Class returnType;
+
+        public NumbaCallableWrapper(PyObject pyObject, Class returnType, List<Class> paramTypes) {
+            super(pyObject);
+            this.returnType = returnType;
+            this.paramTypes = paramTypes;
+        }
+
+        public Class getReturnType() {
+            return returnType;
+        }
+
+        public List<Class> getParamTypes() {
+            return paramTypes;
+        }
+    }
+
+    private static CallableWrapper wrapCallable(PyObject pyObject) {
+        if (pyObject.hasAttribute("ufunc")) {
+            List<PyObject> params = pyObject.getAttribute("types").asList();
+            if (params.isEmpty()) {
+                return new CallableWrapper(pyObject);
+            }
+            String numbaFuncTypes = params.get(0).getStringValue();
+            return parseNumbaVectorized(pyObject, numbaFuncTypes);
+        } else {
+            return new CallableWrapper(pyObject);
+        }
+    }
+
+    private static final Map<Character, Class> numpyType2JavaClass = new HashMap<Character, Class>();
+    {
+        numpyType2JavaClass.put('i', int.class);
+        numpyType2JavaClass.put('l', long.class);
+        numpyType2JavaClass.put('h', short.class);
+        numpyType2JavaClass.put('f', float.class);
+        numpyType2JavaClass.put('d', double.class);
+        numpyType2JavaClass.put('b', byte.class);
+        numpyType2JavaClass.put('?', boolean.class);
+    }
+
+    private static CallableWrapper parseNumbaVectorized(PyObject pyObject, String numbaFuncTypes) {
+        char numpyTypeCode = numbaFuncTypes.charAt(numbaFuncTypes.length() - 1);
+        Class returnType = numpyType2JavaClass.get(numpyTypeCode);
+        if (returnType == null) {
+            return new CallableWrapper(pyObject);
+        }
+
+        List<Class> paramTypes = new ArrayList<>();
+        for (char numpyTypeChar : numbaFuncTypes.toCharArray()) {
+            if (numpyTypeChar != '-') {
```

Comment:
What does this '-' character mean in this context?

Reply:
new comments added to explain the code